broker.go

```diff
@@ -0,0 +1,606 @@
+	lb.outEventsMutex.Lock()
+	defer lb.outEventsMutex.Unlock()
+
+	lb.outEventsChans = append(lb.outEventsChans, ch)
+	cncl := func() {
+		lb.outEventsMutex.Lock()
+		defer lb.outEventsMutex.Unlock()
+		for i, ca := range lb.outEventsChans {
+			if ca == ch {
+				lb.outEventsChans[i] = lb.outEventsChans[len(lb.outEventsChans)-1]
+				lb.outEventsChans[len(lb.outEventsChans)-1] = nil
+				lb.outEventsChans = lb.outEventsChans[:len(lb.outEventsChans)-1]
+				close(ch)
+				break
+			}
+		}
+	}
+	return ch, cncl
+}
+
+// Sync performs a one-off explicit sync with the given peer for a specific CID
+// and updates the latest synced link to it.
+//
+// The Context passed in controls the lifetime of the background sync process,
+// so make sure the context has an appropriate deadline, if any.
+//
+// A SyncFinished channel is returned to allow the caller to wait for this
+// particular sync to complete.  Any OnSyncFinished readers will also get a
+// SyncFinished when the sync succeeds, but only if syncing to the latest using
+// the default selector.
+//
+// It is the responsibility of the caller to make sure the given CID appears
+// after the latest sync in order to avid re-syncing of content that may have
+// previously been synced.
+//
+// The selector sequence, selSec, can optionally be specified to customize the
+// selection sequence during traversal.  If unspecified, the subscriber's
+// default selector sequence is used.
+//
+// Note that the selector sequence is wrapped with a selector logic that will
+// stop traversal when the latest synced link is reached. Therefore, it must
+// only specify the selection sequence itself.
+//
+// See: ExploreRecursiveWithStopNode.
+func (lb *Broker) Sync(ctx context.Context, peerID peer.ID, c cid.Cid, sel ipld.Node, publisher multiaddr.Multiaddr) (<-chan SyncFinished, error) {
+	if peerID == "" {
+		return nil, errors.New("empty peer id")
+	}
+
+	var err error
+	var syncer Syncer
+
+	// If publisher specified, then get URL for http sync, or add multiaddr to peerstore.
+	if publisher != nil {
+		for _, p := range publisher.Protocols() {
+			if p.Code == multiaddr.P_HTTP || p.Code == multiaddr.P_HTTPS {
+				syncer, err = lb.httpSync.NewSyncer(publisher)
+				if err != nil {
+					return nil, err
+				}
+				break
+			}
+		}
+
+		// Not http, so add multiaddr to peerstore.
+		if syncer == nil {
+			peerStore := lb.host.Peerstore()
+			if peerStore != nil {
+				peerStore.AddAddr(peerID, publisher, lb.addrTTL)
+			}
+		}
+	}
+
+	// No syncer yet, so create datatransfer syncer.
+	if syncer == nil {
+		syncer = lb.dtSync.NewSyncer(peerID, lb.topicName)
+	}
+
+	// Start goroutine to get latest root and handle
+	out := make(chan SyncFinished, 1)
+	go func() {
+		defer close(out)
+
+		var updateLatest bool
+		if c == cid.Undef {
+			// Query the peer for the latest CID
+			c, err = syncer.GetHead(ctx)
+			if err != nil {
+				log.Errorw("Cannot get head for sync", "err", err, "peer", peerID)
```

Comment:
nit (feel free to skip), can we use `log.With` above and automatically get the peer id everywhere we log in this func?

Reply:
Fixed